src/vcSHP.cpp

```diff
@@ -0,0 +1,741 @@
+  char *WKTString = nullptr;
+};
+
+void vcSHP_ReleaseRecord(vcSHP_Record &record)
+{
+  switch (record.shapeType)
+  {
+  case vcSHP_Type::Polyline:
+  case vcSHP_Type::Polygon:
+  {
+    udFree(record.data.poly.parts);
+    udFree(record.data.poly.points);
+  }
+  break;
+  case vcSHP_Type::MultiPoint:
+  {
+    udFree(record.data.multiPoint.points);
+  }
+  break;
+  case vcSHP_Type::PolylineZ:
+  case vcSHP_Type::PolygonZ:
+  case vcSHP_Type::MultiPatch:
+  {
+    udFree(record.data.polyZ.parts);
+    udFree(record.data.polyZ.points);
+
+    if (record.data.polyZ.MValue != nullptr)
+      udFree(record.data.polyZ.MValue);
+  }
+  break;
+  case vcSHP_Type::MultiPointZ:
+  {
+    udFree(record.data.multiPointZ.points);
+    if (record.data.multiPointZ.MValue != nullptr)
+      udFree(record.data.multiPointZ.MValue);
+  }
+  break;
+  case vcSHP_Type::PolylineM:
+  case vcSHP_Type::PolygonM:
+  {
+    udFree(record.data.polyM.parts);
+    udFree(record.data.polyM.points);
+
+    if (record.data.polyM.MValue != nullptr)
+      udFree(record.data.polyM.MValue);
+  }
+  break;
+  case vcSHP_Type::MultiPointM:
+  {
+    udFree(record.data.multiPointM.points);
+    if (record.data.multiPointM.MValue != nullptr)
+      udFree(record.data.multiPointM.MValue);
+  }
+  break;
+  default:
+  break;
+  }
+}
+
+
+void vcSHP_Release(vcSHP *pSHP)
+{
+  // release .shp file
+  for (uint32_t i = 0; i < (uint32_t)pSHP->shpRecords.length; ++i)
+    vcSHP_ReleaseRecord(pSHP->shpRecords[i]);
+
+  pSHP->shpRecords.Deinit();
+
+  // release .dbf file
+  vcDBF_Destroy(&pSHP->pDBF);
+
+  // release .prj file
+  if(pSHP->projectionLoad == udR_Success)
+    udFree(pSHP->WKTString);
+}
+
+static void SwapByteOrder(int length, uint8_t *src)
+{
+  for (int i = 0; i < length / 2; i++)
+  {
+    uint8_t temp = src[i];
+    src[i] = src[length - i - 1];
+    src[length - i - 1] = temp;
+  }
+}
+
+static uint32_t UnpackDouble(bool localBigEndian, uint32_t readPos, double *dest, uint8_t *src)
+{
+  if (localBigEndian)
+    SwapByteOrder(8, src + readPos);
+
+  memcpy(dest, src + readPos, 8);
+  return readPos + 8;
+}
+
+static uint32_t UnpackInt32(bool localBigEndian, uint32_t readPos, int32_t *dest, uint8_t *src)
+{
+  if (localBigEndian)
+    SwapByteOrder(4, src + readPos);
+
+  memcpy(dest, src + readPos, 4);
+  return readPos + 4;
+}
+
+udResult vcSHP_LoadShpRecord(vcSHP *pSHP, udFile *pFile, uint32_t &offset, vcSHP_Type defaultShape)
+{
+  udResult result;
+  uint8_t id[4] = {};
+  uint8_t length[4] = {};
+  uint8_t *buffer = nullptr;
+  uint32_t readPosition = 0;
+  int32_t type = 0;
+  int32_t leftBytes = 0;
+
+  vcSHP_Record record;
+  UD_ERROR_CHECK(udFile_Read(pFile, id, 4 * sizeof(uint8_t)));
+  record.id = (id[0] << 24) | (id[1] << 16) | (id[2] << 8) | id[3];
+  offset += 4;
+
+  UD_ERROR_CHECK(udFile_Read(pFile, length, 4 * sizeof(uint8_t)));
+  record.length = (length[0] << 24) | (length[1] << 16) | (length[2] << 8) | length[3];
+  record.length *= 2;
```

Comment:
Might be worth just putting a comment here to indicate that the file stores the length in shorts rather than bytes. Without any comment, this appears to be wrong.

Reply:
fixed.

Reply:
There doesn't seem to be a comment for this.

Reply:
sorry, it was lost after a series of rollback operations, add it again.